src/pr-trigger.ts

```diff
@@ -4,175 +4,147 @@ import { queryPRInfo, deriveStateForPR } from "./pr-info";
+            prHandlers[pr.number] = () => {
+                clearTimeout(timeout);
+                res(true);
             };
-        } else {
-            context.log.error(`No PR with this number exists, (${JSON.stringify(info)})`);
+        });
+        if (aborted) {
+            context.log.info(`Skipped webhook, superseded by a newer one for ${pr.number}`);
             context.res = {
-                status: 422,
-                body: `No PR with this number exists, (${JSON.stringify(info)})`
+                status: 204,
+                body: "NOOPing due to a newer webhook"
             };
+            return;
+        }
+        delete prHandlers[pr.number];
+
+        context.log.info(`Getting info for PR ${pr.number} - ${pr.title || "(title not fetched)"}`);
+
+        // Generate the info for the PR from scratch
+        const info = await queryPRInfo(pr.number);
+        const prInfo = info.data.repository?.pullRequest;
+
+        // If it didn't work, bail early
+        if (!prInfo) {
+            if (event.name === "issue_comment") {
+                context.res = {
+                    status: 204,
+                    body: `NOOPing due to ${pr.number} not being a PR`
+                };
+            } else {
+                context.log.error(`No PR with this number exists, (${JSON.stringify(info)})`);
+                context.res = {
+                    status: 422,
+                    body: `No PR with this number exists, (${JSON.stringify(info)})`
+                };
+            }
+            return;
         }
 
-        return;
-    }
+        // Convert the info to a set of actions for the bot
+        const state = await deriveStateForPR(prInfo);
+        const actions = computeActions(state);
+
+        // Act on the actions
+        await executePrActions(actions, prInfo);
 
-    // Convert the info to a set of actions for the bot
-    const state = await deriveStateForPR(prInfo);
-    const actions = computeActions(state);
+        // We are responding real late in the process, so it might show
+        // as a timeout in GH a few times (e.g. after GH/DT/NPM lookups)
+        context.res = {
+            status: 200,
+            body: actions
+        };
 
-    // Act on the actions
-    await executePrActions(actions, prInfo);
+        function prFromEvent() {
+            switch (event.name) {
+                case "check_suite": {
+                    const [pr] = event.payload.check_suite.pull_requests;
+                    return pr;
+                }
+                case "issue_comment":
+                    return event.payload.issue;
+                case "project_card": {
+                    // Parse project_card.content_url according to
+                    // repository.pulls_url
+                    const url = event.payload.project_card.content_url;
+                    return { number: +url.slice(url.lastIndexOf("/") + 1) };
```

Comment:
`repository.pulls_url` (https://docs.github.com/en/rest/reference/repos#get-a-repository) is a URI Template (https://docs.github.com/en/rest/overview/resources-in-the-rest-api#hypermedia) and `project_card.content_url` is an instance, so I think reversing the template (parsing vs. expanding) is legit?
```JSON
  "pulls_url": "https://api.github.com/repos/DefinitelyTyped/DefinitelyTyped/pulls{/number}",
```

I disliked "hack" because it sounded harsh, for not bothering to dereference the URL, given that the format of the URL is prescribed.

I'd be into checking the whole value matches the template (either the static template `https://api.github.com/repos/DefinitelyTyped/DefinitelyTyped/pulls{/number}` or the `repository.pulls_url` template from the webhook payload) but I don't think it's worth it: my first choice is to assume the payload is valid, as we do in general.

Reply:
I'm just thinking that it won't be hard for GH to tweak them, for example, add a terminating `/` after the number, and this will be problem that will be hard to find.

But if you think that such changes are unlikely enough, then why not:

```suggestion
                    // Parse project_card.content_url according to
                    // repository.pulls_url
                    return { number: +event.payload.project_card.content_url.replace(/^.*\//, "") };
```

and then the whole thing can be converted to a simple arrow function...

Reply:
(Well, almost for the last part...  It runs into the `no-unnecessary-condition` thing again, and I'm beginning to think that it's a bad rule to have...)